### DISCUSSION TITLE
Copy MacPorts UI: List dependencies to be installed before installing them!

### DISCUSSION THREAD

**Discussion by citibeth** [spack/spack#23793] [Ideas]
See example below of what would be nice in Spack.  If you don't like it... you can press Ctrl-C.

Spack already lists out the concretized form. In that listing, it should probably put a star on packages it plans to install.

```
$ port install geeqie
--->  Computing dependencies for geeqie
--->  Dependencies to be installed: exiv2 gnome-themes-standard adwaita-icon-theme gtk3 at-spi2-atk at-spi2-core dbus xorg-libXtst xorg-recordproto libepoxy mesa bison bison-runtime m4 flex indent xorg-dri2proto xorg-glproto xorg-libXxf86vm xorg-xf86vidmodeproto python34 python3_select icon-naming-utils p5.24-xml-simple p5.24-xml-namespacesupport p5.24-xml-sax p5.24-xml-sax-base p5.24-xml-sax-expat librsvg libcroco vala gnome-backgrounds lcms
--->  Fetching distfiles for exiv2
```

**Comment by tgamblin** [comment]
I want this too.  Currently `do_install` just dives in and builds recursively.  It would be better to print some info about the DAG up front and then farm the work out bottom-up like a scheduler would.  We need that for distributed builds anyway (e.g., a SLURM allocation, 1 package per node).

FWIW, Gentoo is pretty good at this, though I think we could be a lot more intuitive:

![gentoo emerge](http://www.krizalys.com/sites/default/files/colors-everywhere.png)

**Comment by adamjstewart** [comment]
Wait, isn't this just `spack spec`? Why not run `spack spec` before each installation if you really want to know what is going to be installed? I guess the only difference would be that `spack spec` doesn't show you what is and isn't already installed. Personally I would really like to see #483 happen.

**Comment by tgamblin** [comment]
#483 would be pretty easy to hack in there.

I like the idea of install also being more informative about what it's doing.  For example, and this is a long-term thing, and not happening this week, but if we have parallel builds in, say, SLURM, I think it would be cool to have something like this:

```
==> Installing dependencies for <pkg>
           Fetch       Building             Fetch Time    Build Time
dep1   [=======------] [ not started     ]  42s           -
dep2   [===----------] [ not started     ]  45s           -
dep3   [=============] [ cab143.llnl.gov ]  48s           3m4s
dep4   [ not started ] [ not started     ]  0s            -
```

So you could see the distributed build happen.

**Comment by tgamblin** [comment]
Note that with the right scheduling you could have fetches happening concurrently with builds even on a single node.

**Comment by alalazo** [comment]
Ok, it seems we have a plan for some fun in the next months :smile:
